build prefs layout from xml and use bauhaus combo

dterrahe opened the pull request:
The preferences dialog is generated from an xml description using an xsl script. There was a bunch of boilerplate for each tab and section. Now these are generated automatically from a section at the start of the xml, which in theory makes it easier for people to customise their own preference dialog (or add a tab with hidden options). But it should generally help maintainance.

Also changed to using bauhaus comboboxes for preference options. With the latest fixes in place I prefer it over the gtk one and the small differences can get annoying when you expect things to work a certain way. This also makes it easier to integrate combos that expose a preference/darktablerc option in a lib module (for example in masks or maps) without breaking cohesion.

Labels in preferences now also support mnemonics. I've added a few for demonstration purposes in the darkroom tab; hold the ALT key to see the first letter of some of the options light up and pressing that letter allows toggling the option or selecting the entry. This could be automated to always pick the first letter (if multiple lines share the same mnemonic it should go around them). But customising might give better usability or this might not appeal at all.

Also fixed a small regression from #11559 that made the "changed" marker in prefs not fixed size, so toggling an option where all others are at default would cause the right columns to move. And corrected some translation markers to not break shortcuts between language switches.

All in all quite a significant change so please help testing.

TurboGit commented:
Anything special to pay attention to? I have tried and and the pref dialog looks ok to me and works fine.

We have a remaining Gtk combo for the 3D-LUT directory selector, which is expected to me.

dterrahe commented:
Whether all changes get saved and activated and correctly show the next time the dialog is opened.

The things that I thought were risky I've tested and seem to work; so please test the things that _I_ think will be fine ;-)

What do you think about adding mnemonics? If we don't want them I should remove the examples from this PR before merging.

dterrahe commented:
Forgot to mention; this also makes columns same size on all (auto-generated) tabs. So the checkboxes for example don't jump left-right when changing between tabs, which to me is more pleasant. Drawback is that on some tabs there's more empty space between description label and option. Could be solved by right-aligning label ;-)

Or reverting to previous behavior. I'm happy to change to whatever you prefer; just wanted to demonstrate this is possible/easy.

TurboGit commented:
> Whether all changes get saved and activated and correctly show the next time the dialog is opened.

Works for me, I didn't test all prefs though :)

TurboGit commented:
> What do you think about adding mnemonics? If we don't want them I should remove the examples from this PR before merging.

I would say that those are really superfluous on the pref dialog. This is really something that we do open with mouse and it seems far more natural to me to change values using the mouse. There is too many items anyway to assign mnemonics to all... All in all I'll remove this from this PR.

TurboGit commented:
> Forgot to mention; this also makes columns same size on all (auto-generated) tabs. So the checkboxes for example don't jump left-right when changing between tabs, which to me is more pleasant. Drawback is that on some tabs there's more empty space between description label and option.

Thinking more about this I would revert to previous behavior. This is because it really depends on the actual translation. For example on the French translation there is some long label which makes the import tab not looking very good:

![image](https://github.com/darktable-org/darktable/assets/467069/4480606c-318f-4330-a568-c86574b152a9)

And we may have similar (or even worst) situation for some other languages.

dterrahe commented:
> This is really something that we do open with mouse

:-) That's a bit chicken/egg; at the moment you _can_ only do it by mouse. With mnemonics and a shortcut, you could easily toggle settings by keyboard only (which could be far quicker if you have the setting in your muscle memory). But obviously, if a setting needs to be toggled that frequently, it should get a separate shortcut anyway...

So all in all, completely agreed and I'll remove this before resubmitting.

dterrahe commented:
- removed mnemonics
- removed column size matching across tabs
- readded support for @capabilties tag in enum options. This is currently only used in security/other/password storage backend to remove libsecret and kwallet from the dropdown if they have not been compiled in.

btw (mostly for fun) we could also do this:
![image](https://github.com/darktable-org/darktable/assets/1549490/edfc93cd-4005-4421-bd4a-678aaa7268b2)
Arguably that makes things more readable, by having the text closer to the option and separating the section titles more. It is also consistent with how we do outlining in dropdowns. And would allow aligning across tabs without creating big gaps in some.

But obviously I don't want to face all the knee-jerk criticism, so let's not.

dterrahe commented:
Now noticing that one (1) of the section headers randomly had a tooltip attached, which is now lost. I could add generic support for a _tooltip_ attribute to sections, but I wonder how discoverable this is if none of the other sections provide a tooltip.

It may be "better" to incorporate this tooltip (for the "shortcuts with multiple instances" section) into the first tooltip of the options in that section, since those tooltips are referring to each other in sequence anyway.

So combine 
"where multiple module instances are present, these preferences control rules that are applied (in order) to decide which module instance shortcuts will be applied to"
and
"if an instance of the module has focus, apply shortcut to that instance\nnote: blending shortcuts always apply to the focused instance"
into, for example,
"where multiple instances of a module are present, apply shortcuts to the instance that has focus\nIf none are focused, the preferences below control rules that are followed (in order) to decide which module instance shortcuts will be applied to.\nnote: blending shortcuts always apply to the focused instance"

What do you think? @elstoc

dterrahe commented:
And yet something else; the "miscellaneous/tags" section's single option seems to be purely concerned with XMPs. Maybe it should be moved to the "storage/XMP" section? Or does it have wider consequences?

elstoc commented:
> What do you think? @elstoc

I'll try to take a look at this change over the weekend

elstoc commented:
IMO if we're going to combine the tooltip options for shortcuts into a single tooltip, it probably needs to be more discoverable. I haven't looked into the detail of what you've done here but is it possible to add a button beside the section header, which shows the tooltip when hovered? 

Also, I'm not entirely convinced about the dismissable message at the top of the shortcuts panel. It goes without saying that the block of prose without capital letters irritates me, but IMO the "click to dismiss permanently" message should be part of the message itself (in bold and/or italic) rather than a tooltip. Perhaps a space between the paragraphs as well?

dterrahe commented:
> is it possible

_Anything_ is possible. As I said above, I can add a "tooltip" attribute to the new XML `section` that allows specifying tooltips for them individually like before.

My "concern" is just that this was only used _once_ which to me indicates it might not be the right approach (because then why not use it more often, which would also increase the chance users would accidentally discover it, on this or another section, and actively go looking for it). The _purpose_ of switching to a generated approach as this PR does is that it makes it much easier to enforce consistency, which leads to predictability and happy users.

Having an additional button _just_ for a tooltip. Sure... Do we have any examples (in dt or elsewhere) where this was deemed the best approach? I don't really care what I'm asked to implement in this case; for once I'm purely on a butt-covering expedition. (Actually, I have to admit that these days I resort to that mode more often than wanting to fight the good fight every single time). How big/what shape\icon should that button be? Shall we involve our resident CSS specialist? Or shall we ask the oracles on pixls? Matrix? Since I _don't_ care, maybe somebody who _does_ can take the lead on this and let me know? I can mark this RFC (for all the good that will do).

But if you asked _me_ (don't!) I'd say since all the individual option checkbuttons in preferences have a tooltip, that's where the user will quickly learn to look. Maybe not the mythical "User" who can't figure this out, but definitely will grasp modules/pipes/histories and is actually _using_ shortcuts, and will someday be the most valued contributor to the project...

So in order to avoid the regression I noticed quite late of dropping the tooltip completely, I suggest for _this_ PR the best way forward would be to merge the tooltips somewhat like I suggested above and then I'll commit to implementing whatever better approach is agreed on later. @TurboGit ?

> It goes without saying that the block of prose without capital letters irritates me, 

I assume you are using en@truecase? In the "english" version we "need" to keep consistently lower case. I'm not sure what happened to the idea of having Truecase as the default. @victoryforce ?

> but IMO the "click to dismiss permanently" message should be part of the message itself (in bold and/or italic) rather than a tooltip. Perhaps a space between the paragraphs as well?

Yes, makes sense; I'll include that in this PR as well. I must have wanted to preserve vertical space, but since this should be dismissed fairly quickly anyway, that's not a real concern. I _am_ a little worried that users might accidentally click on it before having had a chance to read. What are your thoughts on "_click **three times** to dismiss permanently_"?

Any objections against the moves of several options into different sections that I suggested above?

elstoc commented:
> I assume you are using en@truecase? In the "english" version we "need" to keep consistently lower case. 

No. I'm perfectly happy with having buttons, control names etc. in lower case but when it comes to sentences and paragraphs starting with lowercase letters it's just perverse. I could probably get used to en@truecase being the default but I will never get round to rewriting the whole user manual to use it, which is my main argument for keeping the current lower case names as default.

> "miscellaneous/interface/position of scopes module" belongs in "darkroom/general"?

I thought the scopes module was used in other views as well. Is it in the tethering view?

> "miscellaneous/interface/method to use for gettings display profile" -> "processing/image processing"

Not particularly bothered either way here

We might also be getting to the point where splitting some of the panels might be called for. The "darkroom", "processing" and "miscellaneous" sections are all getting quite big.

> merge the tooltips somewhat like I suggested above

Are we talking about _duplicating_ the merged tooltips against each option. That might be more discoverable.

Is there anything I've missed?

dterrahe commented:
> I thought the scopes module was used in other views as well. Is it in the tethering view?

You are right of course; scratch that.

> Are we talking about _duplicating_ the merged tooltips against each option. That might be more discoverable.

That could be the suggested outcome of a wider consultation. For now I want a minimal fix that just doesn't lose the info completely.

> Is there anything I've missed?

This one?
> What are your thoughts on "_click **three times** to dismiss permanently_"?

elstoc commented:
> What are your thoughts on "click three times to dismiss permanently"?

If you like, I don't mind. As long as the useful information is in the user manual as well...

dterrahe commented:
> As long as the useful information is in the user manual as well...

The user manual necessarily provides a lot of detail. So much so that the simple message
"The recommended way to assign shortcuts to visual elements is the [visual shortcut mapping](https://docs.darktable.org/usermanual/development/en/preferences-settings/shortcuts/#visual-shortcut-mapping) mode." might get lost. It could be added to the first paragraph of https://docs.darktable.org/usermanual/development/en/preferences-settings/shortcuts/#defining-shortcuts

dterrahe commented:
> We might also be getting to the point where splitting some of the panels might be called for. The "darkroom", "processing" and "miscellaneous" sections are all getting quite big.

Just wanted to mention that this PR will make this exceedingly easy as creating new sections or even tabs would only require editing `darktableconfig.xml.in`. So after this gets merged anybody can work on this.

So on second thought I'll leave the reorganisation of some options out of this PR.

dterrahe commented:
Added back the info in the "shortcuts with multiple instances" section tooltip to the first option ("focused instance") tooltip.

Also improved the notice at the top of the dialog (and made dismissing it require 3 clicks instead of one accidental one).

And added a (hidden) option to not include the shortcuts tab in the preferences dialog at all, since it can be called up in its own dialog. I may be the only potential user for this option, but I believe it can do no harm.

This is now really ready to go, so that others may then easily reorganise sections/tabs using the new structure introduced here.

TurboGit reviewed (approved):
Works for me, thanks!

dterrahe commented:
Release Note:
- the generation of the preference dialog now takes the specification of tabs and sections from `darktable.xml.in` too, so almost the complete layout can be changed by just editing that file.
- changed the preference dialog dropdowns to use bauhaus widgets so it conforms to (and offers the same behavior) as widgets in the rest of the program.

TurboGit commented:
Thanks!
